xapian-letor/include/xapian-letor/feature.h

```diff
@@ -72,7 +72,6 @@ class XAPIAN_VISIBILITY_DEFAULT Feature {
     /// A bitmask of the statistics this Feature needs.
     stat_flags stats_needed;
 
-  public:
```

Comment:
I think we want to mark the above as `@protected` in the doxygen comments - we're only making them `public` because of limitations from SWIG, not because we want people to treat them as `public` in general.

Also, does all of the above need exposing?  I'd think only the `enum` does.

(And not directly related to this patch, but `stats_needed` should probably be `private` not `protected` - it's readable via public `get_stats()` and can be added to via protected `need_stat()` so nothing should need to directly access it.)

Reply:
> Also, does all of the above need exposing? I'd think only the enum does.
>
I think yes because feature default constructor initializes stats_needed so it has to be public
and making need_stat() protected or private gives swig errors.

Reply:
> Also, does all of the above need exposing? I'd think only the enum does.
>
I think yes because feature default constructor initializes stats_needed so it has to be public
and making need_stat() protected or private gives swig errors.

Reply:
The `Feature` default constructor is a method of the `Feature` class, so should be able to access all members.  The question is whether any SWIG generated code which isn't in a subclass of `Feature` needs to access these things.  In the case of `stats_needed` I can't see why SWIG needs access it directly at all.

Reply:
> The Feature default constructor is a method of the Feature class, so should be able to access all members. 
>
Oh sorry that's not what I meant. I meant that because Feature default constructor is public, we need to make stats_needed public as not doing so, compiler will complain of not seeing stats_needed as it will be defined in private block below, so this is an ordering issue. Also we can't put private parts first as stat_flags need to be defined first in order to define stats_needed.

Reply:
> In the case of stats_needed I can't see why SWIG needs access it directly at all.
>
I think it doesn't need direct access but as I said above we can't make stats_needed private or protected if even that doesn't cause swig errors.

Reply:
> this is an ordering issue

In a C++ class definition you can use a member before it is declared/defined.

Indeed, look at the current code on the branch and you'll see `stats_needed` is **already** defined after the `need_stat()` function refers to it:

```
    void need_stat(stat_flags flag) { 
        stats_needed = stat_flags(stats_needed | flag);
    }                         
                                 
    /// @protected A bitmask of the statistics this Feature needs.
    stat_flags stats_needed;
```

If I move `stats_needed` down to the `private` section and adjust the member initialisation order in the constructor then the code compiles fine.

(It's also possible to use `private:` multiple times, though generally better not to unless there's a good reason to).

Reply:
Ah I see thanks.